transport/redis/handler.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package redis
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/yarpc/internal/errors"
+	"go.uber.org/yarpc/transport"
+	tInternal "go.uber.org/yarpc/transport/internal"
+	"go.uber.org/yarpc/transport/redis/internal"
+)
+
+type handler struct {
+	registry transport.Registry
+	deps     transport.Deps
+
+	server QueueServer
+	stop   chan struct{}
+}
+
+func newHandler(
+	service transport.ServiceDetail,
+	server QueueServer,
+	deps transport.Deps,
+) *handler {
+	return &handler{
+		registry: service.Registry,
+		server:   server,
+		stop:     make(chan struct{}),
+	}
+}
+
+func (h *handler) StartBackground() {
+	go h.start()
+}
+
+func (h *handler) start() {
+	for {
+		select {
+		case <-h.stop:
+			return
+		default:
+			h.handle()
+		}
+	}
+}
+
+func (h *handler) handle() {
+	item, err := h.server.BRPopLPush(time.Second)
+	if err != nil {
+		return
+	}
+
+	start := time.Now()
+	tracer := h.deps.Tracer()
+
+	spanContext, req, err := internal.UnmarshalRPC(tracer, item)
+	if err != nil {
+		h.server.LRem(item)
+		return
+	}
+
+	ctx, span := internal.ExtractOpenTracingSpan(context.Background(), spanContext, req, tracer, "redis", start)
+	defer span.Finish()
+
+	spec, err := h.registry.Choose(ctx, req)
+	if err != nil {
+		internal.UpdateSpanWithErr(span, err)
+		h.server.LRem(item)
+		return
+	}
+
+	if spec.Type() != transport.Oneway {
+		err = errors.UnsupportedTypeError{Transport: "redis", Type: string(spec.Type())}
+		internal.UpdateSpanWithErr(span, err)
+		return
+	}
+
+	tInternal.SafelyCallOnewayHandler(ctx, spec.Oneway(), req)
+	h.server.LRem(item)
+}
+
+func (h *handler) Stop() {
+	h.stop <- struct{}{}
```

Comment:
Close the channel to stop. Not only does that free up resources, it allows
multiple listeners on the channel to wake up rather than just one.